From 62090e89b553a022bf564e76840dc67ea786ec65 Mon Sep 17 00:00:00 2001
From: Nicolas Viennot <Nicolas.Viennot@twosigma.com>
Date: Thu, 28 Oct 2021 21:05:57 +0000
Subject: [PATCH 031/120] net: optimize restore_rule() to not open the
 CR_FD_RULE image file twice

Previously, `open_image(CR_FD_RULE, O_RSTR, pid)` was called twice.
Opening an image file twice is not allowed when streaming the image.
This commit optimizes the code to only open the image file once.

Also improved the error path in restore_ip_dump().

Signed-off-by: Nicolas Viennot <Nicolas.Viennot@twosigma.com>
---
 criu/net.c | 51 ++++++++++++++++++---------------------------------
 1 file changed, 18 insertions(+), 33 deletions(-)

diff --git a/criu/net.c b/criu/net.c
index 7b45f0633..02115c4de 100644
--- a/criu/net.c
+++ b/criu/net.c
@@ -2250,12 +2250,12 @@ static int restore_ip_dump(int type, int pid, char *cmd)
 	sockfd = img_raw_fd(img);
 	if (sockfd < 0) {
 		pr_err("Getting raw FD failed\n");
-		return -1;
+		goto out_image;
 	}
 	tmp_file = tmpfile();
 	if (!tmp_file) {
 		pr_perror("Failed to open tmpfile");
-		return -1;
+		goto out_image;
 	}
 
 	while ((n = read(sockfd, buf, 1024)) > 0) {
@@ -2264,25 +2264,34 @@ static int restore_ip_dump(int type, int pid, char *cmd)
 			pr_perror("Failed to write to tmpfile "
 				  "[written: %d; total: %d]",
 				  written, n);
-			goto close;
+			goto out_tmp_file;
 		}
 	}
 
 	if (fseek(tmp_file, 0, SEEK_SET)) {
 		pr_perror("Failed to set file position to beginning of tmpfile");
-		goto close;
+		goto out_tmp_file;
 	}
 
-	if (img) {
-		ret = run_ip_tool(cmd, "restore", NULL, NULL, fileno(tmp_file), -1, 0);
-		close_image(img);
+	if (type == CR_FD_RULE) {
+		/*
+		 * Delete 3 default rules to prevent duplicates. See kernel's
+		 * function fib_default_rules_init() for the details.
+		 */
+		run_ip_tool("rule", "flush", NULL, NULL, -1, -1, 0);
+		run_ip_tool("rule", "delete", "table", "local", -1, -1, 0);
 	}
 
-close:
+	ret = run_ip_tool(cmd, "restore", NULL, NULL, fileno(tmp_file), -1, 0);
+
+out_tmp_file:
 	if (fclose(tmp_file)) {
 		pr_perror("Failed to close tmpfile");
 	}
 
+out_image:
+	close_image(img);
+
 	return ret;
 }
 
@@ -2304,31 +2313,7 @@ static inline int restore_route(int pid)
 
 static inline int restore_rule(int pid)
 {
-	struct cr_img *img;
-	int ret = 0;
-
-	img = open_image(CR_FD_RULE, O_RSTR, pid);
-	if (!img) {
-		ret = -1;
-		goto out;
-	}
-
-	if (empty_image(img))
-		goto close;
-
-	/*
-	 * Delete 3 default rules to prevent duplicates. See kernel's
-	 * function fib_default_rules_init() for the details.
-	 */
-	run_ip_tool("rule", "flush", NULL, NULL, -1, -1, 0);
-	run_ip_tool("rule", "delete", "table", "local", -1, -1, 0);
-
-	if (restore_ip_dump(CR_FD_RULE, pid, "rule"))
-		ret = -1;
-close:
-	close_image(img);
-out:
-	return ret;
+	return restore_ip_dump(CR_FD_RULE, pid, "rule");
 }
 
 /*
-- 
2.34.1

